### PR TITLE
Cache all Python unpacking if app lastUpdateTime unchanged

### PR DESCRIPTION
**Pull request by paulproteus**
This speeds up the **second** time you start an app.

Before change app first start time: 2.9 sec

```
06-28 22:21:32.940 12868 12868 D MainActivity: onCreate() start
06-28 22:21:35.839 12868 12868 D MainActivity: onCreate() complete
```

Before change app second start time: 2.9 sec

```
06-28 22:22:15.250 12894 12894 D MainActivity: onCreate() start
06-28 22:22:18.214 12894 12894 D MainActivity: onCreate() complete
```

After change app first start time: 3.1 sec

```
06-28 22:19:32.755 12794 12794 D MainActivity: onCreate() start
06-28 22:19:35.879 12794 12794 D MainActivity: onCreate() complete
```

After change app second start time: 1.9 sec

```
06-28 22:20:18.820 12823 12823 D MainActivity: onCreate() start
06-28 22:20:20.889 12823 12823 D MainActivity: onCreate() complete
```

App first start time is the time between "onCreate() start" and "onCreate() complete" being logged when executing the app via `rm -rf android && briefcase update android -rd && briefcase build android && briefcase run android -d @beePhone`.

App **second** start time is the time between "onCreate() start" and "onCreate() complete"  when I quit the app from the Android UI and then click the app icon from the launcher.

I have checked that "onCreate() start" is logged seemingly as soon as I click the app icon and that "onCreate() complete" is logged when the app shows its UI.

Note that even though it looks like this slows down the first launch, I think that was a fluke. The next time I tested it I got 2.8 seconds. This is pretty noisy, I'm afraid.

The use of `lastUpdateTime` scared me at first -- I wanted to use some deterministic APK hash or something. It's the system time at time of APK installation. I think it's completely OK to use it this way. Note that `briefcase run` currently always reinstalls the APK, even if that's not necessary, which means that for people developing an app, they'll never get to experience the speedy start behavior unless they click on the app from their Android device's launcher.

### CODE REVIEW

File: {{ cookiecutter.formal_name }}/app/src/main/java/org/beeware/android/MainActivity.java [hunk 1]
```diff
@@ -92,6 +94,29 @@ private File ensureDir(File dir) throws IOException {
...
+        if (lastUpdateTimeFile.exists()) {
+            BufferedReader reader = new BufferedReader(new InputStreamReader(new
+FileInputStream(lastUpdateTimeFile), StandardCharsets.UTF_8));
+            storedLastUpdateTime = reader.readLine();
+        }
+        try {
+            PackageInfo packageInfo = context.getPackageManager().getPackageInfo(context.getPackageName(), 0);
+            actualLastUpdateTime = String.valueOf(packageInfo.lastUpdateTime);
+        } catch (PackageManager.NameNotFoundException e) {
+            Log.e(TAG, "Unable to find package; using default actualLastUpdateTime: " + actualLastUpdateTime);
```
Comment (freakboy3742) [inline {{ cookiecutter.formal_name }}/app/src/main/java/org/beeware/android/MainActivity.java:113]: Won't the "default" will always be ``null``? If so, why record the value in the log?

File: {{ cookiecutter.formal_name }}/app/src/main/java/org/beeware/android/MainActivity.java [hunk 2]
```diff
@@ -136,6 +161,12 @@ private void unpackPython(Map<String, File> paths) throws IOException {
         File userCodeDir = paths.get("user_code");
         Log.d(TAG, "Unpacking Python assets to base dir " + userCodeDir.getAbsolutePath());
         unpackAssetPrefix(getAssets(), "python", userCodeDir);
+        if (actualLastUpdateTime != null) {
```
Comment (freakboy3742) [inline {{ cookiecutter.formal_name }}/app/src/main/java/org/beeware/android/MainActivity.java:164]: Is this correct? If we've unpacked, it doesn't matter what the actualLastUpdateTime is - we should be updating the storedLastUpdateTime.

File: {{ cookiecutter.formal_name }}/app/src/main/java/org/beeware/android/MainActivity.java (hunk 1 above)
Comment (paulproteus) [reply {{ cookiecutter.formal_name }}/app/src/main/java/org/beeware/android/MainActivity.java:113]: No real reason. I just figured I'd print it for, uh, clarity?

File: {{ cookiecutter.formal_name }}/app/src/main/java/org/beeware/android/MainActivity.java (hunk 2 above)
Comment (paulproteus) [reply {{ cookiecutter.formal_name }}/app/src/main/java/org/beeware/android/MainActivity.java:164]: IMHO if the last update time is null, then I don't want to write it, for two reasons.

1. I have to create a `String` out of `null`, just to learn that it's four bytes long. Ugh, that'd be dumb boilerplate.

2. If it's null, then that means we can't actually compute last update time value, which means that this cache-validation strategy has failed. We can't verify next startup that the app is unchanged. Therefore, I prefer not to write anything to the file.

**Comment by paulproteus** [comment]
My preference here is that we do it the way I have it :) but if you want the changes you suggested, I'm happy to make them my time tomorrow, or have you make them before that and beat me to it. Just trying to pass the baton if possible in the interest of latency-reduction.

File: {{ cookiecutter.formal_name }}/app/src/main/java/org/beeware/android/MainActivity.java (hunk 2 above)
Comment (freakboy3742) [reply {{ cookiecutter.formal_name }}/app/src/main/java/org/beeware/android/MainActivity.java:164]: 👍 That makes sense.

**Pull request merged**